Adding darknet backbone for Yolo-Fastest

aychun opened the pull request:
## Description
Cherry-picked commits to add the Darknet backbone as part of adding support for Yolo-Fastest  (#376) 

We recommend to link at least one existing issue for PR. Before your create a PR, please check if there is an issue for this change.  

## Change(s)

- Add a Darknet class as the backbone for the yolo-fastest model 
- Add a temporary config file for testing the backbone

## Code Formatting

If you PR to either `master` or `dev` branch, you should follow the code linting process. Please check your code with `lint_check.sh` in `./scripts` directory.
For more information, please read the contribution guide in `CONTRIBUTING.md`. 

## Changelog

If your PR is granted to `dev` branch, codeowner will add a brief summary of the change to the **Upcoming Release** section of the [`CHANGELOG.md`](https://github.com/Nota-NetsPresso/netspresso-trainer/blob/master/CHANGELOG.md) file including a link to the PR (formatted in markdown) and a link to your github profile.

For example,

```
- Added a new feature by `@myusername` in [PR 2023](https://github.com/Nota-NetsPresso/netspresso-trainer/pull/2023)
```

illian01 commented on src/netspresso_trainer/models/backbones/experimental/darknet.py:
```diff
@@ -2,18 +2,36 @@
+SUPPORTING_TASK = ["classification", "segmentation", "detection", "pose_estimation"]
+
+
+BLOCK_FROM_LITERAL: Dict[str, Type[nn.Module]] = {
+    "basicblock": BasicBlock,
+    "bottleneck": Bottleneck,
+    "darknetblock": DarknetBlock,
+}
```
I think `BLOCK_FROM_LITERAL` is redundant.
Repeating blocks in YOLO-Fastest is actually `depthwise` version of `DarknetBlock`. If you activate `depthwise` parameter in `DarknetBlock`, This and related lines can be removed.

https://github.com/Nota-NetsPresso/netspresso-trainer/blob/48217e8901efd1cb4ee9851d5bfcafc93ef9845e/src/netspresso_trainer/models/op/custom.py#L608-L631

About `DarknetBlock` with `depthwise`, refer here.

https://github.com/Megvii-BaseDetection/YOLOX/blob/ac58e0a5e68e57454b7b9ac822aced493b553c53/yolox/models/network_blocks.py#L79-L101

illian01 commented on config/model/yolo-fastest/yolo-fastest-tmp.yaml:
```diff
@@ -0,0 +1,114 @@
+          use_act: [True, True, False, False]
+          use_group: [False, False, False, False]
+          num_blocks: [1, 1, 1, 2]
+          hidden_expansion: 4
+        - # stage_3
+          in_channels: [8, 32, 32, 8]
+          out_channels: [32, 32, 8, 8]
+          kernel_sizes: [1, 3, 1, 1]
+          stride: [1, 2, 1, 1]
+          use_act: [True, True, False, False]
+          use_group: [False, True, False, False]
+          num_blocks: [1, 1, 1, 2]
+          hidden_expansion: 6
+        - # stage_4
+          in_channels: [8, 48, 48, 16]
+          out_channels: [48, 48, 16, 16]
+          kernel_sizes: [1, 3, 1, 1]
+          stride: [1, 2, 1, 1]
+          use_act: [True, True, False, False]
+          use_group: [False, True, False, False]
+          num_blocks: [1, 1, 1, 4]
+          hidden_expansion: 6
+        - # stage_5
+          in_channels: [16, 96, 96, 24]
+          out_channels: [96, 96, 24, 24]
+          kernel_sizes: [1, 3, 1, 1]
+          stride: [1, 2, 1, 1]
+          use_act: [True, True, False, False]
+          use_group: [False, True, False, False]
+          num_blocks: [1, 1, 1, 4]
+          hidden_expansion: 5.6666
+        - # stage_6
+          in_channels: [24, 136, 136, 48]
+          out_channels: [136, 136, 48, 48]
+          kernel_sizes: [1, 3, 1, 1]
+          stride: [1, 2, 1, 1]
+          use_act: [True, True, True, False]
+          use_group: [False, True, False, False]
+          num_blocks: [1, 1, 1, 5]
+          hidden_expansion: 4.6666
```
I think `state_params` could be more simlified.
1. I think stem better not be in `stage_params`.
2. I think three blocks of each stages are repeated, and it is similary with `DarknetBlock` without shortcut (if `depthwise` is applied). Can we more simplify configuration?

illian01 reviewed (commented):
Thank you for your effort! 😀 
I made some comment about block patterns. Please check detailed review.

aychun commented:
Thanks for the reviews.  I'll check out the `SeparableConvLayer` you added in the yolo-fastest branch for replacing the Bottleneck block.
I'll request a new review after the updates.

aychun commented:
@illian01 
I removed some redundant fields and also changed the structure to allow different block types for the `stage_stem_block `and `stage_block`. For Yolo-fastest both are darknet blocks with the newly defined `depthwise` option.  I haven't fully figured what type of block should be used for the yolov3 model, but I believe I have generalized the structure enough so that the `Darknet` class can support yolov3 backbone (darknet53) with minimal changes in the future.

illian01 commented on src/netspresso_trainer/models/op/custom.py:
```diff
@@ -613,19 +613,114 @@ def __init__(
-        self.conv2 = ConvLayer(in_channels=hidden_channels, out_channels=out_channels,
-                                kernel_size=3, stride=1, act_type=act_type)
+
+        if is_stem_stage:
+            hidden_channels = round(in_channels * expansion)
+        else:
+            hidden_channels = round(out_channels * expansion)
```
Why is_stem_stage is added here?
Can you explain the purpose of these lines?

aychun replied:
I switch back the round function with int() updated the model config file to make it correctly compute the number of hidden channels.

illian01 commented on src/netspresso_trainer/models/op/custom.py:
```diff
@@ -613,19 +613,114 @@ def __init__(
+        use_act: bool = True,
+        act_type: Optional[str] = None,
+        no_out_act: bool = False,
+        # depthwise_use_act: bool = True,
+        # pointwise_use_act: bool = True,
+        # depthwise_act_type: Optional[str] = None,
+        # pointwise_act_type: Optional[str] = None,
```
These and related lines should be removed

illian01 commented on src/netspresso_trainer/models/backbones/experimental/darknet.py:
```diff
@@ -2,19 +2,36 @@
+SUPPORTING_TASK = ["classification", "segmentation", "detection", "pose_estimation"]
+
+
+BLOCK_FROM_LITERAL: Dict[str, Type[nn.Module]] = {
+    "basicblock": BasicBlock,
+    "bottleneck": Bottleneck,
+    "darknetblock": DarknetBlock,
+}
```
Do we need BasicBlock and Bottleneck for darknet?
I think ConvLayer and DarknetBlock are enough.

illian01 commented on config/model/yolo-fastest/yolo-fastest-tmp.yaml:
I think we better to change this config name to yolo-fastest-detection.yaml.
And how about changing directory name to just yolo

illian01 reviewed (changes requested):
Thank you for your effort! 
I have some minor comments. Please check detailed. 👍

aychun commented:
@illian01 
Fixed the problems and tested with the new YOLOv3FPN module! The code runs without an error but please let me know if there is any room for improvement. Appreciate your help as always :)

illian01 reviewed (approved):
LGTM!
